### ISSUE TITLE
Bug with ClassificationReport / use of classifier?

### ISSUE THREAD

**Issue by jbone** [online-ml/river#504]
## Versions

**river version**:  latest (head)
**Python version**:  3.8
**Operating system**:  Ubuntu 18.04

## Describe the bug

There appears to be a NoneType class maintained / retained in the ClassificationReport’s (or any metrics?) confusion matrix from features fed to metrics update for any None predictions yielded by classifier prior to having enough samples to start feeding it ground truths.

## Steps/code to reproduce

Given a stream of feature dicts, let’s call them frames, denoted f0, f1, f2, ...  We are predicting a target “horizon” (h) frames in the future;  let h=2.

We have a multiclass.onevsone() wrapped around an ALMA classifier.  (The specifics of classifier don’t seem to matter.)  Using defaults for parameters.  The classes will be -1, 0, and 1.

We create a ClassificationReport with defaults.

We proceed to loop through the data calling predict_one on each frame, the first several of which may / will yield None.  Once we are h steps past the first sample f0 (Ie, f0+h) we are able to assess the ground truth of the prior frames that originally yielded none at prediction due to never having seen / learned any ground truth actual classes.

If we update the ClassificationReport at each step, it will “remember” those NoneType predictions as a class and the confusion matrix will include a row/column for the invalid class label value None.

If we subsequently print the ClassificationReport it will generate a stack trace as it tries to sort the classes in order to put the rows and columns in normal order, because NoneType does not support `>` (ie `__gt__`).

Workaround:  don’t update ClassificationReport or metrics using a confusion matrix with prior NoneType predictions made prior to classifiers having seen the actual classes / “burned in” (at least enough to have learned actual classes a certain number times / to have stopped yielding NoneType predictions.)

Better fix:  have confusion matrices ignore any NoneType prediction classes or values.

Apologies for the lack of sample code but since the workaround is pretty clear it’s probably actually an undocumented best practice;  I’m pushing primarily to raise the possible discussion of how to handle the predict / learn flow and metric tracking at “burn-in” and / or see whether somebody says “oh, you’re doing it wrong.”

**Comment by MaxHalford** [comment]
Very interesting, thanks for opening the discussion.

I fully understand the problem you're getting at.

To resume what you've already said, there are possibilities:

1. We trust users to not pass `None` values in their training loop/system. In such a case, the `Metric` is not responsible for filtering out `None`s. In fact, this is exactly what we do in `evaluate.progressive_validation`, as you can see [here](https://github.com/online-ml/river/blob/master/river/evaluate/progressive_validation.py#L49).
2. We smarten `ConfusionMatrix` and make it ignore cases where `y_pred is None`.

Option 1 is always a good idea. We like the idea of responsabilising users and not hiding too much magic behing opaque functions. We really like the fact that users code the training loop themselves. I'm convinced this empowers them.

However, after reading what you've said, I believe it makes enough to sense to add a filter condition to `metrics.ConfusionMatrix`. The latter is the backbone of each classification metric, and thus adding the filter to it will fix this behavior for each classification metric, which is great.

Would you mind issuing a pull request? That way you will get the credit for the change :). If you could add a couple of sentences to explain the fact that `None`s are filtered to handle cold-start cases, that would be the icing on the cake.

**Comment by jbone** [comment]
Will do!

**Comment by jbone** [comment]
What would the preferred interface be?  Presumably this is just an argument to the constructor.  Do we want to simply have a Boolean-valued keyword argument (default to True) eg `... ignore_nones=<True|False>` or perhaps a filter predicate function?

My assumption is the former as the cm implementation is Cython and we probably don’t want to be evaluating Python per se in the middle of that, which an arbitrary filtering function might entail?

Nb. I’ll probably not have a chance to get this into shape for a PR this weekend, but happy to do it!

**Comment by MaxHalford** [comment]
> What would the preferred interface be? Presumably this is just an argument to the constructor. Do we want to simply have a Boolean-valued keyword argument (default to True) eg ... ignore_nones=<True|False> or perhaps a filter predicate function?

I would add this `ignore_none` parameter to the `__init__` of `confusion.ConfusionMatrix`. You would also have to expose the parameter in the constructor of `BinaryMetric` and `MultiClassMetric` which are inherited by the classification metrics.

> My assumption is the former as the cm implementation is Cython and we probably don’t want to be evaluating Python per se in the middle of that, which an arbitrary filtering function might entail?

Correct. A boolean flag that defaults to `True` will do the job.

> Nb. I’ll probably not have a chance to get this into shape for a PR this weekend, but happy to do it!

No pressure! Please do not hesitate to ping me if you struggle.

**Comment by MaxHalford** [comment]
Closing this due to lack of activity! No worries though, we can work on it if the issue pops up again.
